src/Buckle/Buckle/CodeAnalysis/Lowering/Lowerer.cs

```diff
@@ -549,6 +549,33 @@ goto continue
         return base.RewriteMemberAccessExpression(expression);
     }
 
+    protected override BoundExpression RewriteIndexExpression(BoundIndexExpression expression) {
+        /*
+
+        <operand><openBracket><index>]
+
+        ----> <openBracket> is '?['
+
+        (HasValue(<operand>) ? <operand>[<index] : null)
+
+        */
+        if (expression.isNullConditional) {
+            var left = new BoundCallExpression(
+                BuiltinFunctions.HasValue, ImmutableArray.Create<BoundExpression>(expression.operand)
+            );
+
+            var center = new BoundIndexExpression(expression.operand, expression.index, false);
+            var right = new BoundLiteralExpression(null);
+            var op = BoundTernaryOperator.Bind(
+                SyntaxKind.QuestionToken, SyntaxKind.ColonToken, left.type, center.type, right.type
+            );
+
+            return RewriteExpression(new BoundTernaryExpression(left, op, center, right));
+        }
+
+        return base.RewriteIndexExpression(expression);
+    }
```

Comment:
This is the bulk of the logic for this feature, the comment explains how it is being rewritten, similar to `RewriteMemberAccessExpression`.